client/devdocs/docs-selectors/param-type.jsx

```diff
@@ -2,30 +2,87 @@
  * External dependencies
  */
 import React, { PropTypes } from 'react';
-import { get } from 'lodash';
+import { get, has, union } from 'lodash';
+
+/** @type {RegExp} */
+const ARRAY_TYPE_PATTERN = /^Array[.]<([^>]+)>/;
+
+/** @type {RegExp} */
+const OBJECT_TYPE_PATTERN = /^Object[.]<([^,]+), ([^>]+)>/;
 
 /**
- * Matches an expression type
+ * Returns a unified textual description of data types
  *
- * @type {RegExp}
+ * @param {String} name type name from JSDoc output
+ * @returns {String} description of type
  */
-const REGEXP_EXPRESSION_TYPE = /(.*)Type$/;
+const describeType = name => {
+	/** @type {?String[]} */
+	const arrayInfo = name.match( ARRAY_TYPE_PATTERN );
+
+	if ( arrayInfo ) {
+		const [ /* match */, type ] = arrayInfo;
+		return `${ type }[]`;
+	}
+
+	/** @type {?String[]} */
+	const objectInfo = name.match( OBJECT_TYPE_PATTERN );
+
+	if ( objectInfo ) {
+		const [ /* match */, keyType, valueType ] = objectInfo;
+		return `{ ${ keyType }: ${ valueType } }`;
```

Comment:
I will revisit this. I think I agree with you but I didn't touch the React part on my recent work.